Adding optional authentication headers to GET request & config

mio-moto opened the pull request:
This PR implements optional authentication headers. It follows the style of `Seq.App.HttpRequest`. I've omitted optional / additional headers for now.

The version number is not yet bumped, I didn't feel comfortable doing that, you know your versioning better.

I've tested it with a local instance and it works as expected.

Related Issue: #11 Authentication

Commit:
Adding optional authentication headers to GET request & config

liammclennan commented:
Hi @DarkMio,

Thanks for the PR. 

Looking at the code it appears to add an optional 'Authorization' header. Which means that the PR name should perhaps be 'Adding optional Authorization header to GET request & config'. 

The help text, and the property name also don't match the code : "An optional `Name: Value` header, stored as sensitive data, for authentication purposes." since if the user specifies a `Name: Value` pair this will fail.

Rather than change the property name of the help text perhaps change the code that sets the header to allow any header, as seq-app-httprequest does? https://github.com/datalust/seq-app-httprequest/blob/dev/src/Seq.App.HttpRequest/Settings/HeaderSettingFormat.cs

It might also be a good idea to trim the header before assigning it. 

I also have some concern about the authorization header being applied to all targets. For example, someone might use the basic scheme and send their unencrypted credentials to all health check targets. We might need to restrict an app instance to a single target URL if an authentication header is specified. We need to think about this some more... @nblumhardt ?

nblumhardt commented:
Thanks for the ping @liammclennan; I think I'd expect that if I enter an authorization header, and multiple target URLs, then the header would go to all of them. It'll be painful to perform authenticated health checks across a fleet of services if we don't allow this.

nblumhardt commented:
(I also agree that we should replicate the approach for authorization and other headers that Seq.App.HttpRequest uses, if we can 👍 .)

mike-cochrane commented:
Thanks for this PR as it's very timely. I would also vote for allowing other headers as well for extra flexibility.

nblumhardt commented:
Hi @DarkMio - are you keen to push forward with this one, or would you prefer for one of us to pick it up?

mio-moto commented:
Hey hey, sorry for the late reply.

I can do it, I'll just have to see whenever I can move onward with it (schedule is a bit full currently, but expect it in the next few days).
It makes sense to allow key-value header entries - I'll also agree that it makes sense that headers are sent across all URLs in one ingress app (as I can just host multiple apps with similar properties if I need them to differ).

I'll have a deeper look at your attributes and API to render it in the web-ui and will try to keep on the same seq version requirement as the package is on already.

Thanks for the feedback!

Commit:
Adding optional long field with additional header values

mio-moto commented:
I added a few changes. I don't know how you guys handle input checking and all, so I went with a rather simplistic approach for now. Here I'll read the input field, take each line, split on the first double colon and just try to insert it into the request.

Unicode (the unicode encoded in the example below is a 🤔 emoji) causes an exception, duplicated entries will be overwritten, Authorization in the input mask takes precedence over the additional header fields.


UI:
![image](https://user-images.githubusercontent.com/6331542/158792416-3bc5a854-16a4-4575-a045-d55af903ee72.png)

Example:
![image](https://user-images.githubusercontent.com/6331542/158792475-33ef2ab3-81d0-47fc-b9bb-a42f7b9fc3ea.png)


Exception:
![image](https://user-images.githubusercontent.com/6331542/158792737-7ddda57e-bbf8-4a1b-99fc-5660a0915fe1.png)

Commit:
split attributes only once per line, avoid empty entries

nblumhardt commented on src/Seq.Input.HealthCheck/HttpHealthCheck.cs:
```diff
@@ -40,15 +43,32 @@ class HttpHealthCheck
         const string OutcomeSucceeded = "succeeded", OutcomeFailed = "failed";
 
-        public HttpHealthCheck(HttpClient httpClient, string title, string targetUrl, JsonDataExtractor extractor, bool bypassHttpCaching)
+        public HttpHealthCheck(HttpClient httpClient, string title, string targetUrl, string authenticationHeaderValue, string optionalHeader, JsonDataExtractor extractor, bool bypassHttpCaching)
```
The authentication header should be a `Name: Value` pair like the others, it may be e.g. `X-Some-ApiKey`, not necessarily just `Authorization`.

mio-moto replied:
~~I'm not sure how to resolve this here.~~
~~It's sensible that the UI has a specific `Authorization` password field, as I wouldn't want to leak keys or passwords in UI.~~

~~I can either glue both values together, have another optional field to change the authorization attribute name, and/or combine the optionalHeaders with the authorization headers before the constructor.~~

Thanks for you advice!

nblumhardt commented on src/Seq.Input.HealthCheck/HttpHealthCheck.cs:
```diff
@@ -40,15 +43,32 @@ class HttpHealthCheck
+                return;
+            }
+            // the api does not allow overwriting, therefore removal has to happen first.
+            if (request.Headers.Contains(attribute))
```
I think duplicates are in fact allowed in request headers (horrible as that is), so we should probably just pass through each value in turn.

mio-moto replied:
The request API does not allow duplicate header fields, causing an exception as it is a dictionary with an existing key: 
![image](https://user-images.githubusercontent.com/6331542/158959143-114d69bc-cbaa-4a9d-9f94-035820d2443a.png)

(I stumbled upon the behaviour when trying the same thing.)

nblumhardt commented on src/Seq.Input.HealthCheck/HttpHealthCheck.cs:
```diff
@@ -40,15 +43,32 @@ class HttpHealthCheck
             _bypassHttpCaching = bypassHttpCaching;
         }
 
+        private static void AddOrOverwriteHeader(HttpRequestMessage request, string attribute, string value)
```
Nit: codebase doesn't use explicit `private`

nblumhardt replied:
HTTP calls them field "names", may be better to use `name` than `attribute` here

nblumhardt commented on src/Seq.Input.HealthCheck/HealthCheckInput.cs:
```diff
@@ -34,6 +34,26 @@ public class HealthCheckInput : SeqApp, IPublishJson, IDisposable
+            HelpText = "Optional headers sent to each URL in format of `Attribute: Value`.\n" + 
+                       "Headers always need a value, attributes cannot contain spaces, duplicate attributes will be " +
+                       "overwritten in order of appearance.")]
+        public string AdditionalHeader { get; set; }
```
Would prefer to make name (`OtherHeaders`), documentation, and behavior identical to _Seq.App.HttpRequest_ here, if we can.

nblumhardt commented on src/Seq.Input.HealthCheck/HttpHealthCheck.cs:
```diff
@@ -29,6 +30,8 @@ class HttpHealthCheck
         readonly string _title;
         readonly string _targetUrl;
+        readonly string _authenticationHeaderValue;
+        private readonly string _optionalHeader;
```
Nit: codebase doesn't use explicit `private`

nblumhardt commented on src/Seq.Input.HealthCheck/HttpHealthCheck.cs:
```diff
@@ -40,15 +43,32 @@ class HttpHealthCheck
+        private static void AddOrOverwriteHeader(HttpRequestMessage request, string attribute, string value)
+        {
+            // opportunistic abort: header values cannot be empty (https://datatracker.ietf.org/doc/html/rfc7230#section-3.2)
+            if (string.IsNullOrWhiteSpace(value))
```
It's not 100% clear that this is the case, there are rather a lot of threads suggesting that empty field values are acceptable; in the linked EBNF the rule:

```ebnf
field-value    = *( field-content / obs-fold )
```

has a 0-or-more `*` so may be better to just let these through.

mio-moto replied:
You are right, the API does allow _empty_ values, but not _null_ values. I'll add that.

nblumhardt commented on src/Seq.Input.HealthCheck/HttpHealthCheck.cs:
```diff
@@ -72,6 +92,25 @@ public async Task<HealthCheckResult> CheckNow(CancellationToken cancel)
             {
                 var request = new HttpRequestMessage(HttpMethod.Get, probedUrl);
                 request.Headers.Add("X-Correlation-ID", probeId);
+                if(_optionalHeader != null) {
```
Perhaps reusing the code from _Seq.App.HttpRequest_ would help keep complexity down in this class? E.g. https://github.com/datalust/seq-app-httprequest/blob/e23b2328141352f5b4851c7502801f5bfe6511f7/src/Seq.App.HttpRequest/Settings/HeaderSettingFormat.cs#L9 provides request header parsing that handles the two settings accepted by the app.

mio-moto replied:
That is good, however do you suggest copying the code?
I have no problems with that.

nblumhardt reviewed (commented):
Thanks for following up @DarkMio!

Ran through the PR and changes quickly, since we're already maintaining almost identical code in _Seq.App.HttpRequest_, and have some unit tests for various things like header parsing over there, perhaps it would be better in the long run if we "borrow" some code/tests from there?

Commit:
cleaning up explicit private modifiers

Commit:
Combining Auth/Optional Header parsing, adding empty-header support, aligning documentation with Seq.App.HttpRequest

mio-moto commented:
I've updated the PR. I could imagine in the future to merge Seq.App.HttpRequest into one package with two apps (if that is possible) or just have a shared code base. I've introduced some changes to header parsing, due to limitations in the C# Language Level (as I'm not comfortable changing anything in the project from my side, including version numbers and language level)

nblumhardt commented:
Thanks @DarkMio  - I'll merge this through, and consider tweaking the C# language level settings 👍

mio-moto commented:
Very cool! 👍 
Don't forget to increase the version number.

mike-cochrane commented:
Just wondering if there is going to be a new NuGet release soon now that things appear to be ready.

nblumhardt commented:
Hi @mike-cochrane - thanks for the nudge; turns out the repository's NuGet API key had expired; it's updated now, and there's a new `-dev` build on NuGet.org 👍
